source/encoding.c

```diff
@@ -0,0 +1,316 @@
+    size_t remainder_count = (buffer_length % 3);
+    size_t str_index = 0;
+
+    for(size_t i = 0; i < to_encode_len; i += 3 )
+    {
+        uint32_t block = to_encode[ i ];
+
+        block <<= 8;
+        if (i + 1 < buffer_length)
+        {
+            block = block | to_encode[ i + 1 ];
+        }
+
+        block <<= 8;
+        if (i + 2 < to_encode_len)
+        {
+            block = block | to_encode[ i + 2 ];
+        }
+
+        output[str_index++] = base64_encoding_table[(block >> 18) & 0x3F];
+        output[str_index++] = base64_encoding_table[(block >> 12) & 0x3F];
+        output[str_index++] = base64_encoding_table[(block >> 6) & 0x3F];
+        output[str_index++] = base64_encoding_table[block & 0x3F];
+    }
+
+    if(remainder_count > 0)
+    {
+        output[block_count * 4 - 1] = '=';
+        if(remainder_count == 1)
+        {
+            output[block_count * 4 - 2] = '=';
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static void init_decoding_table() {
+    if(!base64_decoding_table_init) {
+        for(size_t i = 0; i < (sizeof(base64_encoding_table) / sizeof(char)); ++i)
+        {
+            size_t index = (size_t)base64_encoding_table[i];
+            base64_decoding_table[index] = (uint8_t)i;
+        }
+
+        base64_decoding_table[(size_t)'='] = base64_sentinal_value;
+        base64_decoding_table_init = 1;
+    }
+}
+
+int aws_base64_decode(const char *to_decode, size_t to_decode_len, uint8_t *output, size_t *output_size) {
+    size_t decoded_length = calculate_base64_decoded_length(to_decode, to_decode_len);
+
+    if(!output) {
+        *output_size = decoded_length;
+        return AWS_OP_SUCCESS;
+    }
+
+    init_decoding_table();
+
+    if (*output_size < decoded_length) {
+        return aws_raise_error(AWS_ERROR_INVALID_BUFFER_SIZE);
+    }
+
+    *output_size = decoded_length;
+
+    size_t block_count = to_decode_len / 4;
+    size_t string_index = 0;
+
+    for(size_t i = 0; i < block_count; ++i)
+    {
+        uint32_t value1 = base64_decoding_table[(size_t)to_decode[string_index++]];
+        uint32_t value2 = base64_decoding_table[(size_t)to_decode[string_index++]];
+        uint32_t value3 = base64_decoding_table[(size_t)to_decode[string_index++]];
+        uint32_t value4 = base64_decoding_table[(size_t)to_decode[string_index++]];
+
+        size_t buffer_index = i * 3;
+        output[buffer_index++] = (uint8_t)((value1 << 2) | ((value2 >> 4) & 0x03));
+        if(value3 != base64_sentinal_value)
+        {
+            output[buffer_index++] = (uint8_t)(((value2 << 4) & 0xF0) | ((value3 >> 2) & 0x0F));
+            if(value4 != base64_sentinal_value)
+            {
+                output[buffer_index] = (uint8_t)((value3 & 0x03) << 6 | value4);
+            }
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+uint8_t *aws_add_uint64_to_buffer(uint8_t *buffer, uint64_t value) {
+    buffer[0] = (uint8_t)(value >> bytes_7) & mask;
```

Comment:
This is the sort of thing that needs optimization.

Here's a portable optimized version for this variant, but we should optimize the rest as well:

```
uint32_t lo = htonl(value);
uint32_t hi = htonl(value >> 32);
memcpy(&buffer[0], &hi, 4);
memcpy(&buffer[0], &lo, 4);
```

It's possible to do slightly better than this on x86_64 by using the 64-bit `BSWAP` instruction.

Reply:
But `htonl` is not portable.
You can avoid the shifting and the bitwise Anding by doing something like this:
```C
uint8_t *write_uint64(uint8_t *buffer, uint64_t value) {
    union {
        uint8_t b0;
        uint8_t b1;
        //....
        uint8_t b7;
        uint64_t all;
    };
    all = value;
    buffer[0] = b7;
    buffer[1] = b6;
    // ...
    buffer[7] = b0;
    return buffer + sizeof(value);
}
```
This is basically 9 assignments.

Reply:
apparently htonl is portable

https://msdn.microsoft.com/en-us/library/windows/desktop/ms740069(v=vs.85).aspx

and it goes without saying it's in posix as well.

Reply:
Worst case we can probe for htonl in CMake; it's a bit of work but reading/writing big-endian values is worth it I think.

Reply:
As far as I can tell, it's supported everywhere that matters.

Reply:
I'll note the header files are different but that's easy enough to deal with.

Reply:
Embedded would be my only real suspect for missing htonl and friends.